src/lib/recurringPurposes.js

```diff
@@ -1,18 +1,248 @@
+        metadata: {
+          addressId
+        }
+      }
+    }
+  }
+  const timeserieWithRel = {
+    ...timeserie,
+    relationships: { ...(timeserie.relationships || {}), ...newRel }
+  }
+
+  return timeserieWithRel
+}
+
+const updateAddressCoordinates = async ({
+  client,
+  newCoordinates,
+  contact,
+  addressId
+}) => {
+  // Query contact to get latest rev
+  const newContact = await queryContactByDocId(client, contact._id)
+  let addressIdx = -1
+  for (let i = 0; newContact.address.length; i++) {
+    if (newContact.address[i].id === addressId) {
+      addressIdx = i
+      break
+    }
+  }
+  if (addressIdx < 0) {
+    log('warn', `No address ${addressId} found on contact ${contact._id}`)
+    return null
+  }
+
+  const currCoordinates = newContact.address[addressIdx].geo?.geo
+  const currCount = newContact.address[addressIdx].geo?.count || 0
+  const currSum = newContact.address[addressIdx].geo?.sum || [0, 0]
+  const newSumLon = currSum[0] + newCoordinates.lon
+  const newSumLat = currSum[1] + newCoordinates.lat
+
+  if (!currCoordinates) {
+    newContact.address[addressIdx].geo.geo = [
+      newCoordinates.lon,
+      newCoordinates.lat
+    ]
+  } else {
+    // XXX - This is an incremental average of longitudes and latitudes, that does not
+    // take into consideration of curvature of the Earth. Hence, this is an approximation.
+    // We consider this is acceptable as the coordinates are supposed to be very close, i.e. few meters.
+    // For more precise computations, one should typically use a geodesic center of all the points.
+
+    const count = currCount + 1
+    const avgLon = newSumLon / count
+    const avgLat = newSumLat / count
+
+    newContact.address[addressIdx].geo.geo = [avgLon, avgLat]
+    log('info', `New address coordinates set : ${[avgLon, avgLat]}`)
+    log(
+      'info',
+      `Distance from previous coordinates: ${geodesicDistance(
+        { lon: currCoordinates[0], lat: currCoordinates[1] },
+        { lon: newCoordinates.lon, lat: newCoordinates.lat }
+      )}`
+    )
+  }
+
+  newContact.address[addressIdx].geo.count = currCount + 1
+  newContact.address[addressIdx].geo.sum = [newSumLon, newSumLat]
+
+  log(
+    'info',
+    `Update contact ${contact.displayName} on address ${contact.address[addressIdx].formattedAddress}`
+  )
+  return client.save(newContact)
+}
+
+export const findStartAndEnd = (timeserie, contacts) => {
+  const { closestStart, closestEnd } = findClosestStartAndEnd(
+    timeserie,
+    contacts
+  )
+  let matchingStart = null,
+    matchingEnd = null
+  if (closestStart?.distance < MAX_SPATIAL_THRESHOLD_M) {
+    matchingStart = closestStart
+  }
+  if (closestEnd?.distance < MAX_SPATIAL_THRESHOLD_M) {
+    matchingEnd = closestEnd
+  }
+  return { matchingStart, matchingEnd }
+}
+
+export const shouldSetCommutePurpose = (start, end) => {
+  if (!start || !end) {
+    return false
+  }
+  const startCategory = start.address?.geo?.cozyCategory
```

Comment:
this `cozyCategory` can't be written by the user?

Reply:
Not directly at least. This is set through the modal associating an address to a contact ;  if the user changes the address category, it will change the `cozyCategory`